File: parser.py
from parsec import many1 as one_or_more, digit, string, try_choice, joint

natural_number = one_or_more(digit())
negative_number = joint(string("-"), natural_number)
integer = try_choice(negative_number, natural_number)
real_number = joint(integer, string("."), natural_number)
number = try_choice(real_number, integer)

# >>> number.parse("-12")
# ('-', ['1', '2'])

File: parser.hs
import Control.Applicative (Alternative (some, (<|>)))
import Text.Parsec.Char (digit, string)
import Text.Parsec.String (Parser)

oneOrMore = some

joint p1 p2 = (++) <$> p1 <*> p2

naturalNumber :: Parser String
naturalNumber = oneOrMore digit

negativeNumber = string "-" `joint` naturalNumber

integer = negativeNumber <|> naturalNumber

realNumber = integer `joint` string "." `joint` naturalNumber

number = realNumber <|> integer